### PR TITLE
[Experimental] Add support for production browser source maps

### PR DESCRIPTION
**Pull request by timneutkens**
Adds support for enabling sourcemaps for the browser in production (development already has sourcemaps).

```js
module.exports = {
  experimental: {
    productionBrowserSourceMaps: true
  }
}
```

Initial implementation for #7579

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 

**test/integration/production-browser-sourcemaps/test/index.test.js** 

<details>
<summary>Expand output</summary>

  ● Test suite failed to run

    ReferenceError: jasmine is not defined

       6 | import { recursiveReadDir } from 'next/dist/lib/recursive-readdir'
       7 | 
    >  8 | jasmine.DEFAULT_TIMEOUT_INTERVAL = 1000 * 60 * 2
         | ^
       9 | const appDir = join(__dirname, '../')
      10 | const nextConfig = join(appDir, 'next.config.js')
      11 | let nextConfigContent

      at Object.<anonymous> (integration/production-browser-sourcemaps/test/index.test.js:8:1)

</details>

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Server Mode</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | zeit/next.js canary  | timneutkens/next.js add/production-sourcemaps-flag | Change |
| - | - | - | - |
| buildDuration | 13.3s | 13.6s | ⚠️ +324ms |
| nodeModulesSize | 58.5 MB | 58.5 MB | ⚠️ +485 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | zeit/next.js canary  | timneutkens/next.js add/production-sourcemaps-flag | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.265 | 2.322 | ⚠️ +0.06 |
| / avg req/sec | 1103.98 | 1076.69 | -27.29 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.521 | 1.562 | ⚠️ +0.04 |
| /error-in-render avg req/sec | 1643.97 | 1600.65 | -43.32 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/production-sourcemaps-flag | Change |
| - | - | - | - |
| main-HASH.js gzip | 6.67 kB | 6.67 kB | ✓ |
| webpack-HASH.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..e49d.js gzip | 10.4 kB | 10.4 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 57 kB | 57 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/production-sourcemaps-flag | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 5.71 kB | 5.71 kB | ✓ |
| webpack-HASH..dule.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..dule.js gzip | 6.85 kB | 6.85 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 52.4 kB | 52.4 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/production-sourcemaps-flag | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 26.3 kB | 26.3 kB | ✓ |
| Overall change | 26.3 kB | 26.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/production-sourcemaps-flag | Change |
| - | - | - | - |
| _app.js gzip | 1.26 kB | 1.26 kB | ✓ |
| _error.js gzip | 3.28 kB | 3.28 kB | ✓ |
| hooks.js gzip | 881 B | 881 B | ✓ |
| index.js gzip | 222 B | 222 B | ✓ |
| link.js gzip | 2.06 kB | 2.06 kB | ✓ |
| routerDirect.js gzip | 279 B | 279 B | ✓ |
| withRouter.js gzip | 278 B | 278 B | ✓ |
| Overall change | 8.26 kB | 8.26 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/production-sourcemaps-flag | Change |
| - | - | - | - |
| _app.module.js gzip | 604 B | 604 B | ✓ |
| _error.module.js gzip | 2.09 kB | 2.09 kB | ✓ |
| hooks.module.js gzip | 383 B | 383 B | ✓ |
| index.module.js gzip | 223 B | 223 B | ✓ |
| link.module.js gzip | 1.52 kB | 1.52 kB | ✓ |
| routerDirect..dule.js gzip | 281 B | 281 B | ✓ |
| withRouter.m..dule.js gzip | 278 B | 278 B | ✓ |
| Overall change | 5.38 kB | 5.38 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/production-sourcemaps-flag | Change |
| - | - | - | - |
| _buildManifest.js gzip | 61 B | 61 B | ✓ |
| _buildManife..dule.js gzip | 61 B | 61 B | ✓ |
| Overall change | 122 B | 122 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/production-sourcemaps-flag | Change |
| - | - | - | - |
| index.html gzip | 926 B | 926 B | ✓ |
| link.html gzip | 936 B | 936 B | ✓ |
| withRouter.html gzip | 923 B | 923 B | ✓ |
| Overall change | 2.79 kB | 2.79 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Serverless Mode</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | zeit/next.js canary  | timneutkens/next.js add/production-sourcemaps-flag | Change |
| - | - | - | - |
| buildDuration | 14.7s | 14.5s | -144ms |
| nodeModulesSize | 58.5 MB | 58.5 MB | ⚠️ +485 B |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/production-sourcemaps-flag | Change |
| - | - | - | - |
| main-HASH.js gzip | 6.67 kB | 6.67 kB | ✓ |
| webpack-HASH.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..e49d.js gzip | 10.4 kB | 10.4 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 57 kB | 57 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/production-sourcemaps-flag | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 5.71 kB | 5.71 kB | ✓ |
| webpack-HASH..dule.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..dule.js gzip | 6.85 kB | 6.85 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 52.4 kB | 52.4 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/production-sourcemaps-flag | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 26.3 kB | 26.3 kB | ✓ |
| Overall change | 26.3 kB | 26.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/production-sourcemaps-flag | Change |
| - | - | - | - |
| _app.js gzip | 1.26 kB | 1.26 kB | ✓ |
| _error.js gzip | 3.28 kB | 3.28 kB | ✓ |
| hooks.js gzip | 881 B | 881 B | ✓ |
| index.js gzip | 222 B | 222 B | ✓ |
| link.js gzip | 2.06 kB | 2.06 kB | ✓ |
| routerDirect.js gzip | 279 B | 279 B | ✓ |
| withRouter.js gzip | 278 B | 278 B | ✓ |
| Overall change | 8.26 kB | 8.26 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/production-sourcemaps-flag | Change |
| - | - | - | - |
| _app.module.js gzip | 604 B | 604 B | ✓ |
| _error.module.js gzip | 2.09 kB | 2.09 kB | ✓ |
| hooks.module.js gzip | 383 B | 383 B | ✓ |
| index.module.js gzip | 223 B | 223 B | ✓ |
| link.module.js gzip | 1.52 kB | 1.52 kB | ✓ |
| routerDirect..dule.js gzip | 281 B | 281 B | ✓ |
| withRouter.m..dule.js gzip | 278 B | 278 B | ✓ |
| Overall change | 5.38 kB | 5.38 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/production-sourcemaps-flag | Change |
| - | - | - | - |
| _buildManifest.js gzip | 61 B | 61 B | ✓ |
| _buildManife..dule.js gzip | 61 B | 61 B | ✓ |
| Overall change | 122 B | 122 B | ✓ |

</details>

<details>
<summary><strong>Serverless bundles</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/production-sourcemaps-flag | Change |
| - | - | - | - |
| _error.js | 850 kB | 850 kB | ✓ |
| 404.html | 4.18 kB | 4.18 kB | ✓ |
| hooks.html | 3.81 kB | 3.81 kB | ✓ |
| index.js | 851 kB | 851 kB | ✓ |
| link.js | 888 kB | 888 kB | ✓ |
| routerDirect.js | 880 kB | 880 kB | ✓ |
| withRouter.js | 880 kB | 880 kB | ✓ |
| Overall change | 4.36 MB | 4.36 MB | ✓ |

</details>

</details>

File: packages/next/build/webpack-config.ts [hunk 1]
```diff
@@ -643,6 +645,7 @@ export default async function getBaseWebpackConfig(
           cache: path.join(distDir, 'cache', 'next-minifier'),
           parallel: config.experimental.cpus || true,
           terserOptions,
+          sourceMap: productionBrowserSourceMaps,
```
Comment (Timer) [inline packages/next/build/webpack-config.ts:648]:
```suggestion
```

https://github.com/webpack-contrib/terser-webpack-plugin/blob/10d64125165af65a37a22dd83df84f7904850617/src/index.js#L527-L542

**Review by Timer — CHANGES_REQUESTED** [review]

**Comment by ijjk** [comment]
## Failing test suites 

**test/integration/production-browser-sourcemaps/test/index.test.js** 

<details>
<summary>Expand output</summary>

  ● Test suite failed to run

    ReferenceError: jasmine is not defined

       6 | import { recursiveReadDir } from 'next/dist/lib/recursive-readdir'
       7 | 
    >  8 | jasmine.DEFAULT_TIMEOUT_INTERVAL = 1000 * 60 * 2
         | ^
       9 | const appDir = join(__dirname, '../')
      10 | const nextConfig = join(appDir, 'next.config.js')
      11 | let nextConfigContent

      at Object.<anonymous> (integration/production-browser-sourcemaps/test/index.test.js:8:1)

</details>

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Server Mode</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | zeit/next.js canary  | timneutkens/next.js add/production-sourcemaps-flag | Change |
| - | - | - | - |
| buildDuration | 11.1s | 10.5s | -683ms |
| nodeModulesSize | 58.5 MB | 58.5 MB | ⚠️ +472 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | zeit/next.js canary  | timneutkens/next.js add/production-sourcemaps-flag | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 1.784 | 1.712 | -0.07 |
| / avg req/sec | 1401.72 | 1459.91 | ⚠️ +58.19 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.071 | 0.975 | -0.1 |
| /error-in-render avg req/sec | 2333.92 | 2564.92 | ⚠️ +231 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/production-sourcemaps-flag | Change |
| - | - | - | - |
| main-HASH.js gzip | 6.67 kB | 6.67 kB | ✓ |
| webpack-HASH.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..e49d.js gzip | 10.4 kB | 10.4 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 57 kB | 57 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/production-sourcemaps-flag | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 5.71 kB | 5.71 kB | ✓ |
| webpack-HASH..dule.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..dule.js gzip | 6.85 kB | 6.85 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 52.4 kB | 52.4 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/production-sourcemaps-flag | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 26.3 kB | 26.3 kB | ✓ |
| Overall change | 26.3 kB | 26.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/production-sourcemaps-flag | Change |
| - | - | - | - |
| _app.js gzip | 1.26 kB | 1.26 kB | ✓ |
| _error.js gzip | 3.28 kB | 3.28 kB | ✓ |
| hooks.js gzip | 881 B | 881 B | ✓ |
| index.js gzip | 222 B | 222 B | ✓ |
| link.js gzip | 2.06 kB | 2.06 kB | ✓ |
| routerDirect.js gzip | 279 B | 279 B | ✓ |
| withRouter.js gzip | 278 B | 278 B | ✓ |
| Overall change | 8.26 kB | 8.26 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/production-sourcemaps-flag | Change |
| - | - | - | - |
| _app.module.js gzip | 604 B | 604 B | ✓ |
| _error.module.js gzip | 2.09 kB | 2.09 kB | ✓ |
| hooks.module.js gzip | 383 B | 383 B | ✓ |
| index.module.js gzip | 223 B | 223 B | ✓ |
| link.module.js gzip | 1.52 kB | 1.52 kB | ✓ |
| routerDirect..dule.js gzip | 281 B | 281 B | ✓ |
| withRouter.m..dule.js gzip | 278 B | 278 B | ✓ |
| Overall change | 5.38 kB | 5.38 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/production-sourcemaps-flag | Change |
| - | - | - | - |
| _buildManifest.js gzip | 61 B | 61 B | ✓ |
| _buildManife..dule.js gzip | 61 B | 61 B | ✓ |
| Overall change | 122 B | 122 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/production-sourcemaps-flag | Change |
| - | - | - | - |
| index.html gzip | 926 B | 926 B | ✓ |
| link.html gzip | 936 B | 936 B | ✓ |
| withRouter.html gzip | 923 B | 923 B | ✓ |
| Overall change | 2.79 kB | 2.79 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Serverless Mode</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | zeit/next.js canary  | timneutkens/next.js add/production-sourcemaps-flag | Change |
| - | - | - | - |
| buildDuration | 11.2s | 11.6s | ⚠️ +419ms |
| nodeModulesSize | 58.5 MB | 58.5 MB | ⚠️ +472 B |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/production-sourcemaps-flag | Change |
| - | - | - | - |
| main-HASH.js gzip | 6.67 kB | 6.67 kB | ✓ |
| webpack-HASH.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..e49d.js gzip | 10.4 kB | 10.4 kB | ✓ |
| framework.HASH.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 57 kB | 57 kB | ✓ |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons) Modern</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/production-sourcemaps-flag | Change |
| - | - | - | - |
| main-HASH.module.js gzip | 5.71 kB | 5.71 kB | ✓ |
| webpack-HASH..dule.js gzip | 746 B | 746 B | ✓ |
| de003c3a9d30..dule.js gzip | 6.85 kB | 6.85 kB | ✓ |
| framework.HA..dule.js gzip | 39.1 kB | 39.1 kB | ✓ |
| Overall change | 52.4 kB | 52.4 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/production-sourcemaps-flag | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 26.3 kB | 26.3 kB | ✓ |
| Overall change | 26.3 kB | 26.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/production-sourcemaps-flag | Change |
| - | - | - | - |
| _app.js gzip | 1.26 kB | 1.26 kB | ✓ |
| _error.js gzip | 3.28 kB | 3.28 kB | ✓ |
| hooks.js gzip | 881 B | 881 B | ✓ |
| index.js gzip | 222 B | 222 B | ✓ |
| link.js gzip | 2.06 kB | 2.06 kB | ✓ |
| routerDirect.js gzip | 279 B | 279 B | ✓ |
| withRouter.js gzip | 278 B | 278 B | ✓ |
| Overall change | 8.26 kB | 8.26 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages Modern</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/production-sourcemaps-flag | Change |
| - | - | - | - |
| _app.module.js gzip | 604 B | 604 B | ✓ |
| _error.module.js gzip | 2.09 kB | 2.09 kB | ✓ |
| hooks.module.js gzip | 383 B | 383 B | ✓ |
| index.module.js gzip | 223 B | 223 B | ✓ |
| link.module.js gzip | 1.52 kB | 1.52 kB | ✓ |
| routerDirect..dule.js gzip | 281 B | 281 B | ✓ |
| withRouter.m..dule.js gzip | 278 B | 278 B | ✓ |
| Overall change | 5.38 kB | 5.38 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/production-sourcemaps-flag | Change |
| - | - | - | - |
| _buildManifest.js gzip | 61 B | 61 B | ✓ |
| _buildManife..dule.js gzip | 61 B | 61 B | ✓ |
| Overall change | 122 B | 122 B | ✓ |

</details>

<details>
<summary><strong>Serverless bundles</strong></summary>

|  | zeit/next.js canary  | timneutkens/next.js add/production-sourcemaps-flag | Change |
| - | - | - | - |
| _error.js | 850 kB | 850 kB | ✓ |
| 404.html | 4.18 kB | 4.18 kB | ✓ |
| hooks.html | 3.81 kB | 3.81 kB | ✓ |
| index.js | 851 kB | 851 kB | ✓ |
| link.js | 888 kB | 888 kB | ✓ |
| routerDirect.js | 880 kB | 880 kB | ✓ |
| withRouter.js | 880 kB | 880 kB | ✓ |
| Overall change | 4.36 MB | 4.36 MB | ✓ |

</details>

</details>

File: test/integration/production-browser-sourcemaps/test/index.test.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
+/* eslint-env jest */
+/* global jasmine */
```
Comment (Timer) [inline test/integration/production-browser-sourcemaps/test/index.test.js:2]:
```suggestion
```

File: test/integration/production-browser-sourcemaps/test/index.test.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+import fs from 'fs-extra'
+import { join } from 'path'
+import { nextBuild } from 'next-test-utils'
+import { recursiveReadDir } from 'next/dist/lib/recursive-readdir'
+
+jasmine.DEFAULT_TIMEOUT_INTERVAL = 1000 * 60 * 2
```
Comment (Timer) [inline test/integration/production-browser-sourcemaps/test/index.test.js:8]:
```suggestion
jest.setTimeout(1000 * 60 * 2)

```

**Review by Timer — CHANGES_REQUESTED** [review]

**Review by Timer — APPROVED** [review]

**Comment by sawyerh** [comment]
Thank you 🙏

**Pull request merged**